store/smt/store.go

```diff
@@ -0,0 +1,211 @@
+package smt
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"io"
+	"sync"
+	"time"
+
+	dbm "github.com/cosmos/cosmos-sdk/db"
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/tracekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/lazyledger/smt"
+)
+
+var (
+	_ types.KVStore                 = (*Store)(nil)
+	_ types.CommitStore             = (*Store)(nil)
+	_ types.CommitKVStore           = (*Store)(nil)
+	_ types.Queryable               = (*Store)(nil)
+	_ types.StoreWithInitialVersion = (*Store)(nil)
+)
+
+var (
+	prefixLen      = 1
+	versionsPrefix = []byte{0}
+	dataPrefix     = []byte{1}
+	indexPrefix    = []byte{2}
+	afterIndex     = []byte{3}
+)
+
+// Store Implements types.KVStore and CommitKVStore.
+type Store struct {
+	tree *smt.SparseMerkleTree
+	db   dbm.DBReadWriter
+
+	version int64
+
+	opts struct {
```

Comment:
This will actually be removed when we get to the KV store implementation. The SMT can be simplified since it replicates the functionality of the decoupled store.

Reply:
I'm going to actually rebase to remove the SMT store from this PR, since it's not actually relevant until we get to the KV store stage.